### ISSUE TITLE
The behaviour of ``Client.wait()`` needs better documentation when ``timeout`` is used.

### ISSUE THREAD

**Issue by delcypher** [docker/docker-py#618]
Based on reading the documentation for [``Client.wait()``](http://docker-py.readthedocs.org/en/latest/api/#wait) I expected ``-1`` to be returned if the specified ``timeout`` was hit. Instead an exception (``requests.exceptions.ReadTimeout``) seems to be hit in different ways internally depending on whether I'm using Python2 or Python3.

Here is an example

```python
import docker

def main():
  client = docker.Client()

  container = client.create_container(
    image='ubuntu:14.04',
    command=['sleep', '100']
    )

  try:
    client.start(container=container['Id'])
    # Now wait
    exitCode = client.wait(container=container['Id'], timeout=5)
    if exitCode == -1:
      print("Timeout??")
  finally:
    client.remove_container(container=container['Id'], force=True)

if __name__ == '__main__':
  main()
```

**Python2 response**

```
Traceback (most recent call last):
  File "./unclear-timeout.py", line 21, in <module>
    main()
  File "./unclear-timeout.py", line 14, in main
    exitCode = client.wait(container=container['Id'], timeout=5)
  File "/usr/lib/python2.7/site-packages/docker/utils/decorators.py", line 15, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/docker/client.py", line 1123, in wait
    res = self._post(url, timeout=timeout)
  File "/usr/lib/python2.7/site-packages/docker/client.py", line 103, in _post
    return self.post(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 508, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 433, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: UnixHTTPConnectionPool(host='localhost', port=None): Read timed out. (read timeout=5)
```

**Python3 response**

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 372, in _make_request
    httplib_response = conn.getresponse(buffering=True)
TypeError: getresponse() got an unexpected keyword argument 'buffering'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 374, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.4/http/client.py", line 1171, in getresponse
    response.begin()
  File "/usr/lib/python3.4/http/client.py", line 351, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.4/http/client.py", line 313, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.4/socket.py", line 374, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 597, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/util/retry.py", line 245, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/packages/six.py", line 310, in reraise
    raise value
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 544, in urlopen
    body=body, headers=headers)
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 376, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/usr/lib/python3.4/site-packages/requests/packages/urllib3/connectionpool.py", line 304, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
requests.packages.urllib3.exceptions.ReadTimeoutError: UnixHTTPConnectionPool(host='localhost', port=None): Read timed out. (read timeout=5)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./unclear-timeout.py", line 21, in <module>
    main()
  File "./unclear-timeout.py", line 14, in main
    exitCode = client.wait(container=container['Id'], timeout=5)
  File "/usr/lib/python3.4/site-packages/docker/utils/decorators.py", line 15, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/usr/lib/python3.4/site-packages/docker/client.py", line 1123, in wait
    res = self._post(url, timeout=timeout)
  File "/usr/lib/python3.4/site-packages/docker/client.py", line 103, in _post
    return self.post(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python3.4/site-packages/requests/sessions.py", line 508, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/lib/python3.4/site-packages/requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3.4/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3.4/site-packages/requests/adapters.py", line 433, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: UnixHTTPConnectionPool(host='localhost', port=None): Read timed out. (read timeout=5)
```

**Thoughts**

* If returning ``-1`` is not for when the ``timeout`` is hit, when will it be returned?
* Right now ``requests.exceptions.ReadTimeout`` being thrown looks like an internal implementation detail of ``docker-py`` that a client should **not** need to know. I think ``docker-py`` should be throwing it's own exception for this case and should document the exception that will be thrown.

**Comment by shin-** [comment]
Hm, alright, I've had some time to look into this. On your thoughts:

- I feel the docs are quite specific about when `-1` will be returned,

> Returns the value `-1` if no `StatusCode` is returned by the API.

I guess we could rephrase into

> Returns the value `-1` if the API responds without a `StatusCode` attribute.

- This is interesting, because right now this exception will be returned any time the specified HTTP timeout is reached, it applies across the board, The fact that you seem to point it out as a `Client.wait` specific problem indicates that there may be a debate to be had about having this particular method behave differently, depending on whether we value being explicit over being consistent. Also it's a breaking change for people depending on the current behavior. @mpetazzoni @aanand, thoughts on that point?

**Comment by delcypher** [comment]
> Returns the value -1 if the API responds without a StatusCode attribute.

This phrasing seems better to me.

> This is interesting, because right now this exception will be returned any time the specified HTTP timeout is reached, it applies across the board, The fact that you seem to point it out as a Client.wait specific problem indicates that there may be a debate to be had about having this particular method behave differently, depending on whether we value being explicit over being consistent. Also it's a breaking change for people depending on the current behavior.

You don't necessarily have to break the interface. You could continue to raise the ``requests.exceptions.ReadTimeout`` exception but this needs to be explicitly documented and not changed. I personally don't like the fact you are leaking internal implementation details but it may be to late to change that.

**Comment by aanand** [comment]
I'm in favour of the wording change suggested by @shin-. As for `ReadTimeout`, I'm undecided, but I don't think we should let that discussion hold up the 1.3.0 release.

Also, what's up with the exception at the top of the Python 3 output?

    TypeError: getresponse() got an unexpected keyword argument 'buffering'

**Comment by shin-** [comment]
Fixed in 69ef9e5668254adf8c951571660b11d4c89b2529
